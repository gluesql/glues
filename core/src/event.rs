use {
    crate::{
        data::{Directory, Note},
        types::DirectoryId,
    },
    strum_macros::Display,
};

#[derive(Clone, Debug, Display)]
pub enum Event {
    #[strum(to_string = "Key::{0}")]
    Key(KeyEvent),

    #[strum(to_string = "Entry::{0}")]
    Entry(EntryEvent),

    #[strum(to_string = "Notebook::{0}")]
    Notebook(NotebookEvent),

    Cancel,
}

#[derive(Clone, Debug, Display)]
pub enum EntryEvent {
    OpenMemory,
    OpenCsv(String),
    OpenJson(String),
    OpenFile(String),
    OpenGit {
        path: String,
        remote: String,
        branch: String,
    },
    OpenMongo {
        conn_str: String,
        db_name: String,
    },
}

#[derive(Clone, Debug, Display)]
pub enum NotebookEvent {
    OpenDirectory(DirectoryId),
    CloseDirectory(DirectoryId),

    SelectNote(Note),
    SelectDirectory(Directory),

    RenameNote(String),
    RenameDirectory(String),

    RemoveNote,
    RemoveDirectory,

    ShowNoteActionsDialog,
    CloseNoteActionsDialog,

    ShowDirectoryActionsDialog,
    CloseDirectoryActionsDialog,

    AddNote(String),
    AddDirectory(String),

    OpenNote,
    EditNote,
    ViewNote,
    BrowseNoteTree,

    UpdateNoteContent(String),

    CloseEntryDialog,
}

#[derive(Clone, Copy, Debug, Display)]
pub enum KeyEvent {
    B,
    E,
    H,
    J,
    K,
    L,
    M,
    O,
    Num(NumKey),
    Left,
    Right,
    Esc,
}

#[derive(Clone, Copy, Debug, Display)]
pub enum NumKey {
    One,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
    Zero,
}

impl From<EntryEvent> for Event {
    fn from(event: EntryEvent) -> Self {
        Self::Entry(event)
    }
}

impl From<NotebookEvent> for Event {
    fn from(event: NotebookEvent) -> Self {
        Self::Notebook(event)
    }
}

impl From<KeyEvent> for Event {
    fn from(event: KeyEvent) -> Self {
        Self::Key(event)
    }
}

impl From<NumKey> for KeyEvent {
    fn from(num_key: NumKey) -> Self {
        KeyEvent::Num(num_key)
    }
}

impl From<NumKey> for usize {
    fn from(num_key: NumKey) -> Self {
        match num_key {
            NumKey::One => 1,
            NumKey::Two => 2,
            NumKey::Three => 3,
            NumKey::Four => 4,
            NumKey::Five => 5,
            NumKey::Six => 6,
            NumKey::Seven => 7,
            NumKey::Eight => 8,
            NumKey::Nine => 9,
            NumKey::Zero => 0,
        }
    }
}

impl std::ops::Add<usize> for NumKey {
    type Output = usize;

    fn add(self, rhs: usize) -> Self::Output {
        let n = usize::from(self).saturating_add(rhs);

        if n > u16::MAX as usize {
            u16::MAX as usize
        } else {
            n
        }
    }
}
