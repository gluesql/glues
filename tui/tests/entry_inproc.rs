mod common;
use common::{AppTestExt as _, TerminalTestExt as _};

use color_eyre::Result;
use ratatui::crossterm::event::KeyCode;

#[tokio::test]
async fn entry_nav_enter_opens_instant_inproc() -> Result<()> {
    let (mut app, mut term) = common::setup_app_and_term().await?;

    // initial draw (home)
    app.draw_frame(&mut term)?;

    // move down then up, then Enter to open Instant via selection
    app.press('j').await;
    app.press('k').await;
    app.key(KeyCode::Enter).await;

    // draw and assert notebook content is visible (e.g., sample note)
    app.draw_frame(&mut term)?;
    term.assert_contains("Sample Note");

    Ok(())
}

#[tokio::test]
async fn entry_quit_with_q_inproc() -> Result<()> {
    let (mut app, mut term) = common::setup_app_and_term().await?;
    app.draw_frame(&mut term)?;

    let quit = app.press('q').await;
    assert!(quit);

    Ok(())
}

#[tokio::test]
async fn entry_help_overlay_open_close_inproc() -> Result<()> {
    let (mut app, mut term) = common::setup_app_and_term().await?;
    app.draw_frame(&mut term)?;

    // open help (currently bound to 'a')
    app.press('a').await;
    app.draw_frame(&mut term)?;
    term.assert_contains("Press any key to close");

    // any key closes help
    app.press('x').await;
    app.draw_frame(&mut term)?;
    term.assert_not_contains("Press any key to close");

    Ok(())
}
