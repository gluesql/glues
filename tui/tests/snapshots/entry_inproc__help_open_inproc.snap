---
source: tui/tests/entry_inproc.rs
expression: text
snapshot_kind: text
---
 Glues - TUI note-taking app offering complete data control and flexible storage options               [?] Show keymap 
                                                                                                                        
                                                                                                                        
┌─────────────────────────────────────────────────────────Help─────────────────────────────────────────────────────────┐
│                                                                                                                      │
│  Glues offers various storage options to suit your needs:                                                            │
│                                                                                                                      │
│  Instant                                                                                                             │
│  Data is stored in memory and only persists while the app is running.                                                │
│  This option is useful for testing or temporary notes as it is entirely volatile.                                    │
│                                                                                                                      │
│  Local                                                                                                               │
│  Notes are stored locally as separate files.                                                                         │
│  This is the default option for users who prefer a simple, file-based approach without any remote synchronization.   │
│                                                                                                                      │
│  Git                                                                                                                 │
│  Git storage requires three inputs: `path`, `remote`, and `branch`.                                                  │
│  The `path` should point to an existing local Git repository, similar to the file storage path.                      │
│  For example, you can clone a GitHub repository and use that path.                                                   │
│  The `remote` and `branch` specify the target remote repository and branch for synchronization.                      │
│  When you modify notes or directories, Glues will automatically sync changes with the specified remote repository.   │
│                                                                                                                      │
│  MongoDB                                                                                                             │
│  MongoDB storage allows you to store your notes in a MongoDB database, providing a scalable and centralized          │
│  solution for managing your notes.                                                                                   │
│  You need to provide the MongoDB connection string and the database name.                                            │
│  Glues will handle storing and retrieving notes from the specified database.                                         │
│  This option is ideal for users who prefer a centralized storage solution or need robust, reliable data storage.     │
│                                                                                                                      │
│  CSV or JSON                                                                                                         │
│  These formats store notes as simple log files, ideal for quick data exports or reading logs.                        │
│  CSV saves data in comma-separated format, while JSON uses JSONL (JSON Lines) format.                                │
│                                                                                                                      │
│                                                                                                                      │
│                                                Press any key to close                                                │
│                                                                                                                      │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
